### PR TITLE
[FLINK-21899][table] Introduce SOURCE_WATERMARK built-in function to preserve watermark from source

### PR DESCRIPTION
**Pull request by wuchong**
## What is the purpose of the change

The `SOURCE_WATERMARK` function doesn't have concrete implementation. The eval() function should throw a meaningful exception message to indicate users the function should only be used in DDL to generate a watermark preserved from source system.

## Brief change log

- Add `SOURCE_WATERMARK` `BuiltInFunctionDefinition`.
- Add `SourceWatermarkFunction` scalar function which throws exception in `eval()` method. 

## Verifying this change

- Add tests to verify the exception in `TableSourceITCase`. 

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by wuchong** [comment]
cc @leonardBang

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/runtime/stream/sql/TableSourceITCase.scala [hunk 1]
```diff
@@ -19,13 +19,15 @@
...
+import org.apache.flink.core.testutils.FlinkMatchers.containsCause
+import org.apache.flink.table.api.TableException
 import org.apache.flink.table.api.bridge.scala._
 import org.apache.flink.table.planner.factories.TestValuesTableFactory
 import org.apache.flink.table.planner.runtime.utils.{StreamingTestBase, TestData, TestingAppendSink, TestingRetractSink}
 import org.apache.flink.table.planner.utils._
+import org.apache.flink.table.runtime.functions.scalar.SourceWatermarkFunction
 import org.apache.flink.table.utils.LegacyRowResource
 import org.apache.flink.types.Row
-
```
Comment (leonardBang) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/runtime/stream/sql/TableSourceITCase.scala:28]: minor: unnecessary change

**Review by leonardBang — APPROVED** [review]
LGTM, one minor  check style I think you can fix when merge
